### PR TITLE
docs: add guide to adding a new expression

### PR DESCRIPTION
**Pull request by tshauck**
## Which issue does this PR close?

Part of #370 (haven't added anything on cast or aggregate, mainly scalar func)

## Rationale for this change

Make it easier for people to contribute new expressions.

## What changes are included in this PR?

A new markdown file in the contributor guide.

## How are these changes tested?

~Not yet, I haven't tried building the docs. Will do before marking this PR ready.~

Built the docs and reviewed the HTML.

### CODE REVIEW

**Comment by tshauck** [comment]
Notes from comet community meeting:

* Possible to get link for presentation slides and video?
  * Andy to publish video
  * Liang Chi has slides here: https://docs.google.com/presentation/d/1H0fF2MOkkBK8fPBlnqK6LejUeLcVD917JhVWfp3mb8A/edit#slide=id.p
* `CometSparkSessionExtension` injects rules into spark session extension
  * E.g. replace spark scan with comet scan
  * Replace spark execution with comet execution class
* `QueryPlanSerde` takes spark expression an serializes via proto
* Cast/types support is not fully supported (e.g. `Add(left, right, _)` checks if `supportedDataTypes(left.dataType)`.
* operator.proto defines `Operator`
* `message Expr`... need to update `expr.proto` (See `Add` for example)
  * left, right, return type
* `PhysicalPlanner` is the physical planner struct
  * `create_expr` and `create_plan` are two most important parts
  * gets children from protobuf message and calls `create_expr` or `create_plan` to generate the physical plan
  * `spark_expr.expr_struct` is matched on then passed to appropriate part
  * Can sometimes use built-in support for expressions (e.g. is not null expression), this _only_ works if the implement is compatible w/ spark (e.g. cast)
* Adding a Cast
  * datafusion's `PhysicalExpr` is implemented for `Cast`.
* Spark version differences (e.g. `failOnError` is a common change from 3.3 to 3.4+)
  * use shim layer to handle API differences
* `CometExpressionSuite` is for testing
  * use `checkSparkAnswerAndOperator`


Questions:

* Which expressions to support? Try to get support for the full expression.
  * 3.4 main difference is failOnError
  * Expect 4.0 to be reasonably difference
* How to handle new expressions? Add new expression into datafusion if _very_ common expression with standard behavior
* Which other areas are important? Aiming for complete coverage... `CAST` is recently a hotbed of activity and high priority... tests are important
* Unrelated to expressions: first release? Andy has started copying over the basic datafusion scripts to get out a source release, but ultimate goal is jar file with native execution.

**Comment by andygrove** [comment]
@tshauck video link is https://drive.google.com/file/d/1POU4lFAZfYwZR8zV1X2eoLiAmc1GDtSP/view?usp=sharing

**Comment by tshauck** [comment]
I think this is ready for review. The first draft I wrote was pretty slanted towards scalar function expressions since that's what I did w/ unhex, but after the meeting on Wed, I made some updates so hopefully it's more general. I'm not 100% sure on the things I haven't touched yet, so certainly open to feedback on any of it.

Here's what it looks like included in the docs...

image

File: docs/source/contributor-guide/adding_a_new_expression.md [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+You may have a specific expression in mind that you'd like to add, but if not, you can review the [expression coverage document](https://github.com/apache/datafusion-comet/blob/f08fcadd5fbdb5b04293d33e654f6c16f81b70c4/doc/spark_builtin_expr_coverage.txt) to see which expressions are not yet supported.
+
+## Adding the Expression
+
+Once you have the expression you'd like to add, you should take inventory of the following:
+
+1. What is the Spark expressions behavior across different Spark versions? These make good test cases, and will inform you of any compatibility issues such as an API change that will have to be addressed.
+2. Check if the expression is already implemented in DataFusion and if the is compatible with the Spark expression.
+   1. If it is, you can potentially reuse the existing implementation though you'll need to add tests to verify compatibility.
+   2. If it's not, consider an initial version in DataFusion Comet and potentially backport it into DataFusion if the expression would be supported by it.
```
[the remark this answers is not in the document]
Comment (tshauck) [reply docs/source/contributor-guide/adding_a_new_expression.md]: 👍 -- https://github.com/apache/datafusion-comet/pull/422/commits/5dfe312d7ceccf4851035aec062d5458b35b19df

File: docs/source/contributor-guide/adding_a_new_expression.md [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+  val failOnErrorExpr = exprToProtoInternal(unHex._2, inputs)
+
+  val optExpr =
+    scalarExprToProtoWithReturnType("unhex", e.dataType, childExpr, failOnErrorExpr)
+  optExprWithInfo(optExpr, expr, unHex._1)
+```
+
+A few things to note here:
+
+* The `isSpark32` check is used to fall back to Spark's implementation of `unhex` in Spark 3.2, as only versions after that have the `failOnError` parameter.
```
[the remark this answers is not in the document]
Comment (tshauck) [reply docs/source/contributor-guide/adding_a_new_expression.md]: Thanks, better description here: https://github.com/apache/datafusion-comet/pull/422/commits/a599c172c80ebfaa6d026fde2bba1a75f4849ed5... related to https://issues.apache.org/jira/browse/SPARK-40924.

**Comment by andygrove** [comment]
This is looking great @tshauck. Could you fix the merge conflict?

**Comment by tshauck** [comment]
Thanks, @andygrove -- I think the conflict is resolved now.

File: docs/source/index.rst [hunk 3]
```diff
@@ -58,7 +58,11 @@ as a native runtime to achieve improvement in terms of query efficiency and quer
    Comet Plugin Overview <contributor-guide/plugin_overview>
    Development Guide <contributor-guide/development>
    Debugging Guide <contributor-guide/debugging>
+<<<<<<< HEAD
```
Comment (kazuyukitanimura) [inline docs/source/index.rst:61]: Looks like this is a merge/rebase error

**Review by kazuyukitanimura — APPROVED** [review]
LGTM pending CI

File: docs/source/index.rst (hunk 3 above)
Comment (tshauck) [reply docs/source/index.rst]: Ah, thanks! [1c37b49](https://github.com/apache/datafusion-comet/pull/422/commits/1c37b490b5e0d80d6de5b3ff6844f489a87777cf)

**Comment by tshauck** [comment]
@andygrove Conflict should actually be fixed now, thanks @kazuyukitanimura

**Review by andygrove — APPROVED** [review]
LGTM. Thank you @tshauck

**Pull request merged**